includes/wp-admin/import/class-mastodon.php

```diff
@@ -0,0 +1,412 @@
+		$error_message = \__( 'Sorry, there has been an error.', 'activitypub' );
+		$file          = \get_attached_file( self::$import_id );
+
+		\WP_Filesystem();
+
+		global $wp_filesystem;
+		$import_folder = $wp_filesystem->wp_content_dir() . 'import/';
+		self::$archive = $import_folder . \basename( \basename( $file, '.txt' ), '.zip' );
+
+		// Clean up working directory.
+		if ( $wp_filesystem->is_dir( self::$archive ) ) {
+			$wp_filesystem->delete( self::$archive, true );
+		}
+
+		// Unzip package to working directory.
+		\unzip_file( $file, self::$archive );
+		$files = $wp_filesystem->dirlist( self::$archive );
+
+		if ( ! isset( $files['outbox.json'] ) ) {
+			echo '<p><strong>' . \esc_html( $error_message ) . '</strong><br />';
+			echo \esc_html__( 'The archive does not contain an Outbox file, please try again.', 'activitypub' ) . '</p>';
+		}
+
+		// phpcs:ignore WordPress.WP.AlternativeFunctions.file_get_contents_file_get_contents
+		self::$outbox = \json_decode( \file_get_contents( self::$archive . '/outbox.json' ) );
+
+		\wp_suspend_cache_invalidation();
+		\wp_defer_term_counting( true );
+		\wp_defer_comment_counting( true );
+
+		/**
+		 * Fires when the Mastodon import starts.
+		 */
+		\do_action( 'import_start' );
+
+		$result = self::import_posts();
+
+		\wp_suspend_cache_invalidation( false );
+		\wp_defer_term_counting( false );
+		\wp_defer_comment_counting( false );
+
+		$wp_filesystem->delete( $import_folder, true );
+		\wp_import_cleanup( self::$import_id );
+
+		if ( \is_wp_error( $result ) ) {
+			echo '<p><strong>' . \esc_html( $error_message ) . '</strong><br />';
+			echo \esc_html( $result->get_error_message() ) . '</p>';
+		} else {
+			echo '<p>';
+			/* translators: Home URL */
+			\printf( \wp_kses_post( \__( 'All done. <a href="%s">Have fun!</a>', 'activitypub' ) ), \esc_url( \admin_url() ) );
+			echo '</p>';
+		}
+
+		/**
+		 * Fires when the Mastodon import ends.
+		 */
+		\do_action( 'import_end' );
+	}
+
+	/**
+	 * Process posts.
+	 *
+	 * @return true|\WP_Error True on success, WP_Error on failure.
+	 */
+	public static function import_posts() {
+		// phpcs:ignore WordPress.NamingConventions.ValidVariableName.UsedPropertyNotSnakeCase
+		foreach ( self::$outbox->orderedItems as $post ) {
+			// Skip boosts.
+			if ( 'Announce' === $post->type ) {
+				continue;
+			}
+
+			if ( ! is_activity_public( \get_object_vars( $post ) ) ) {
+				continue;
+			}
+
+			// @todo: Skip replies to comments and import them as comments.
+
+			$post_data = array(
+				'post_author'  => self::$author,
+				'post_date'    => $post->published,
+				'post_excerpt' => $post->object->summary ?? '',
+				'post_content' => $post->object->content,
+				'post_status'  => 'publish',
+				'post_type'    => 'post',
+				'tags_input'   => \array_map(
+					function ( $tag ) {
+						if ( 'Hashtag' === $tag->type ) {
+							return \ltrim( $tag->name, '#' );
+						}
+
+						return '';
+					},
+					$post->object->tag
+				),
```

Comment:
What about adding an indicator that it is/was an import, like we have with comments?

```PHP
'meta_input' => array(
	'protocol' => 'activitypub',
	'source_id' => 'https://mastodon.social/...'
)
```

So that it is able to query these posts and to `(C)RUD` them later on?

Reply:
I've added `_source_id` d5e633d45524e503e0131d85659a22348f1674e6 for now. Not sure that's really needed?

We'd want any meta key to be "private" (prefixed with an underscore) and ideally they should also be registered I guess. I've skipped the registration for now, since we don't expect it to be used/updated/managed.